Only set is encrypted when encryption is enabled

nickvergessen opened the pull request:
Smashbox shareDir fails since https://github.com/owncloud/core/pull/17540

The reason is we check the file cache now and the file cache is set to "file is encrypted" when a file is uploaded to a mount point, when the file lives in a mount point.
### TODO
- [x] Add unit test
- [x] Add update step to fix all files when encryption is off and oC === 8.1.0
- [x] Fill placeholder issue with details: https://github.com/owncloud/core/issues/17846
### Steps to reproduce

**Encryption must be DISABLED**
1. Share a folder
2. As recipient edit a file in the folder
3. As owner open the edited file
### Testing matrix

| Instance | Reset | Test |
| --- | --- | --- |
| Install of 8.1.0 | :+1: | :white_check_mark: |
| Install of 8.1.0 + Enable Enc | :no_entry: | :white_check_mark: |
| Install of 8.0 + Update to 8.1.0 + Update to branch | :+1: | :white_check_mark: |
| Install of 8.0 + Enable Enc + Update to 8.1.0 + Update to branch | :warning: Log | :white_check_mark: |
| Install of 8.0 + Enable Enc + Update to 8.1.0 + Enable new Encryption  + Update to branch | :no_entry: | :white_check_mark: |
| Install of 8.0 + Enable/Disable Enc + Update to 8.1.0 + Update to branch | :no_entry: | :white_check_mark: |
| Install of 8.0 + Enable/Disable Enc + Decrypt all + Update to 8.1.0 + Update to branch | :+1: | :white_check_mark: |

@schiesbn please help with tests
cc @DeepDiver1975 thats why the smashbox tests fail on jenkins without enc and locking

nickvergessen committed:
Only set is encrypted when encryption is enabled

nickvergessen committed:
Add a unit test for the disabled encryption case

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/f70c9584-9e9c-4abb-a7f9-36b9e12b58e0) was created.

nickvergessen commented:
@DeepDiver1975 where do update files for core go?

nickvergessen committed:
Add an update script to reset the value

In case encryption was not enabled, we accidently set encrypted = 1 for
files inside mount points, since 8.1.0. This breaks opening the files in
8.1.1 because we fixed the code that checks if a file is encrypted.
In order to fix the file, we need to reset the flag of the file. However,
the flag might be set because the file is in fact encrypted because it was
uploaded at a time where encryption was enabled.

So we can only do this when:
- Current version of ownCloud before the update is 8.1.0 or 8.2.0.(0-2)
- Encryption is disabled
- files_encryption is not known in the app config

If the first two are not the case, we are save. However, if files_encryption
values exist in the config, we might have a false negative here.
Now if there is no file with unencrypted size greater 0, that means there are
no files that are still encrypted with "files_encryption" encryption. So we
can also safely reset the flag here.

If this is not the case, we go with "better save then sorry" and don't change
the flag but write a message to the ownCloud log file.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/d16f6063-403a-456c-a7ef-86549f642633) was created.

nickvergessen commented:
@rperezb please have a look, this needs testing before 8.1.1 goes out (especially the update script)

nickvergessen commented on apps/files/appinfo/update.php:
```diff
@@ -0,0 +1,96 @@
+	)) {
+
+	// Encryption is not enabled
+	if (!\OC::$server->getEncryptionManager()->isEnabled()) {
```
Is this enough @schiesbn 
Or should we check the config for the value and appconfig for `encryption` app?

schiesbn replied:
isEnabled() already checks it: https://github.com/owncloud/core/blob/master/lib/private/encryption/manager.php#L78

nickvergessen commented:
@schiesbn can you confirm the following testing matrix:

### Testing matrix

Instance | Reset | Test
-------- | ----- | ----
Install of 8.1.0 | :+1: | :white_check_mark: 
Install of 8.1.0 + Enable Enc | :no_entry: |
Install of 8.0 | :+1: |
Install of 8.0 + Enable Enc | :no_entry: |
Install of 8.0 + Enable/Disable Enc | :no_entry: |
Install of 8.0 + Enable/Disable Enc + Decrypt all | :+1: |

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/0f25afcd-e7fa-4b5d-a6aa-29516b59b3df) was created.

nickvergessen commented:
@owncloud-bot retest this please

schiesbn commented:
> @schiesbn can you confirm the following testing matrix

Test matrix looks good

PVince81 commented:
I wasn't able to reproduce the original issue with shared folders.

However the only case where I could reproduce it is as follows:
1. Leave encryption **disabled**
2. Mount an SFTP external storage under "/sftp"
3. With cadaver: "put test.txt"
4. With cadaver: "cp test.txt sftp/test.txt".

This does a Webdav COPY and uses the `copyFromStorage` code path.
In this case the target file "sftp/test.txt" receives the bogus encrypted flag.

I'll check if the migration fixes that.

PVince81 commented:
:+1: migration fixed the flag and on this branch the flag is not set any more.
After enabling encryption, the flag is still properly set on regular and external storage.

DeepDiver1975 commented on apps/files/appinfo/update.php:
```diff
@@ -0,0 +1,96 @@
+ * @param \OCP\IDBConnection $conn
+ */
+function owncloud_reset_encrypted_flag(\OCP\IDBConnection $conn) {
+	$conn->executeUpdate('UPDATE `*PREFIX*filecache` SET `encrypted` = 0');
```
can we add a condition to this to reduce the record sets being touched?

DeepDiver1975 commented on apps/files/appinfo/update.php:
```diff
@@ -0,0 +1,96 @@
+		if (empty($oldEncryption)) {
+			owncloud_reset_encrypted_flag($conn);
+		} else {
+			$query = $conn->prepare('SELECT * FROM `*PREFIX*filecache` WHERE `encrypted` = 1 AND `unencrypted_size` > 0', 1);
```
use select count - no need to get all the record sets

schiesbn replied:
we just get the first one (see limit), think this should be OK

DeepDiver1975 replied:
Ah - there is a limit involved - hard to see :see_no_evil:

Commit:
add condition to update query

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/8990972d-b113-49a7-89d2-2d40d1de7886) was created.

schiesbn commented:
added a additional condition to update only db entries with encrypted=1.

@DeepDiver1975 Please have a final look I would like to merge & backport before RC1 so that it get tested by QA and some other people before we release it.

schiesbn commented:
@owncloud-bot retest this please

DeepDiver1975 commented:
:+1:

DeepDiver1975 commented:
I'll create the backport pr ....

schiesbn commented:
Thanks for taking care about the backport!
